macros 1.1: Allow proc_macro functions to declare attributes to be mark as used

keeperofdakeys opened the pull request:
This PR allows proc macro functions to declare attribute names that should be marked as used when attached to the deriving item. There are a few questions for this PR.

- Currently this uses a separate attribute named `#[proc_macro_attributes(..)]`, is this the best choice?
- In order to make this work, the `check_attribute` function had to be modified to not error on attributes marked as used. This is a pretty large change in semantics, is there a better way to do this?
- I've got a few clones where I don't know if I need them (like turning `item` into a `TokenStream`), can these be avoided?
- Is switching to `MultiItemDecorator` the right thing here?

Also fixes https://github.com/rust-lang/rust/issues/37563.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @arielb1 (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

alexcrichton commented on src/libsyntax/feature_gate.rs:
```diff
@@ -820,7 +826,7 @@ impl<'a> Context<'a> {
             // before the plugin attributes are registered
             // so we skip this then
-            if !is_macro {
+            if !is_macro && !attr::is_used(attr) {
```
How come this part was modified as well? Presumably this means that we're not gating unused attributes now?

keeperofdakeys replied:
Using mark_used appears to only disable the unused_attribute lint, not the custom_attributes error. So without this change the compiler will still give the custom attribute error for used attributes. The only alternative I can see is registering the attribute name in plugin_attributes.

jseyfried replied:
@alexcrichton It means we're not `custom_attribute`-checking *used* attributes, which should be fine because unknown/custom attributes should never be `mark_used`.

jseyfried replied:
@alexcrichton It means we're not `custom_attribute`-checking *used* attributes, which should be fine because unknown/custom attributes should never be `mark_used`.

nrc replied:
"because unknown/custom attributes should never be mark_used" - where is this enforced?

alexcrichton commented on src/libsyntax_ext/proc_macro_registrar.rs:
```diff
@@ -170,11 +171,39 @@ impl<'a> Visitor for CollectCustomDerives<'a> {
+        let proc_attrs: Vec<_> = item.attrs.iter()
+            .filter(|a| a.check_name("proc_macro_attributes"))
+            .filter_map(|a|
+              a.meta_item_list().or_else(|| {
```
This spacing/indentation here is a bit difficult to read, could this be updated to follow the surrounding code?

keeperofdakeys commented:
Thanks for the feedback, I'll work on the tests and get back to you.

jseyfried commented on src/librustc_metadata/creader.rs:
```diff
@@ -624,8 +624,12 @@ impl<'a> CrateLoader<'a> {
-                let derive = SyntaxExtension::CustomDerive(Box::new(CustomDerive::new(expand)));
+                                      expand: fn(TokenStream) -> TokenStream,
+                                      attributes: &[&'static str]) {
+                let attrs: Vec<_> = attributes.iter().map(|s| InternedString::new(s)).collect();
```
nit: `.map(|s| InternedString::new(s))` -> `.map(InternedString::new)`
Also, it `InternedString` is going to become `Name` soon, so it would be better for `attrs` to be `Vec<Name>` instead of `Vec<InternedString>`. In that case, you would use  `.map(token::intern)`.

jseyfried commented on src/librustc_metadata/creader.rs:
```diff
@@ -624,8 +624,12 @@ impl<'a> CrateLoader<'a> {
+                                      attributes: &[&'static str]) {
+                let attrs: Vec<_> = attributes.iter().map(|s| InternedString::new(s)).collect();
+                let derive = SyntaxExtension::CustomDerive(
+                    Box::new(CustomDerive::new(expand, &attrs))
```
This should pass `attrs` directly instead of by reference so that `CustomDerive::new` doesn't need to clone.

jseyfried commented on src/libsyntax/ext/base.rs:
```diff
@@ -509,7 +509,7 @@ pub enum SyntaxExtension {
     IdentTT(Box<IdentMacroExpander>, Option<Span>, bool),
 
-    CustomDerive(Box<MultiItemModifier>),
+    CustomDerive(Box<MultiItemDecorator>),
```
This should stay `MultiItemModifier`.

nrc replied:
Why? I'd have thought Decorator is more appropriate?

jseyfried commented on src/libsyntax/feature_gate.rs:
```diff
@@ -638,6 +638,12 @@ pub const KNOWN_ATTRIBUTES: &'static [(&'static str, AttributeType, AttributeGat
+                                            "proc_macro",
+                                            "the `#[proc_macro_attributes]` attribute \
+                                             is an experimental feature",
+                                            cfg_fn!(proc_macro))),
```
We don't need to add a new kind of attribute -- extending `proc_macro_derive` is fine.

jseyfried commented on src/libsyntax_ext/deriving/custom.rs:
```diff
@@ -12,66 +12,76 @@ use std::panic;
-use syntax::print::pprust;
+use syntax::parse::token::InternedString;
+
+fn mark_attrs_used<I: Iterator<Item=Attribute>>(attrs: I, whitelist: &[InternedString]) {
```
This should be a `Visitor`.

jseyfried replied:
e.g. something like
```rust
struct MarkAttrs<'a>(&'a [InternedString]);

impl<'a> Visitor for MarkAttrs<'a> {
    fn visit_attribute(&mut self, attr: &Attribute) {
        if self.0.contains(attr.name()) {
            attr::mark_used(attr);
        }
    }
}
```

jseyfried commented on src/libsyntax_ext/deriving/custom.rs:
```diff
@@ -12,66 +12,76 @@ use std::panic;
-    pub fn new(inner: fn(TokenStream) -> TokenStream) -> CustomDerive {
-        CustomDerive { inner: inner }
+    pub fn new(inner: fn(TokenStream) -> TokenStream,
+               attrs: &[InternedString])
```
This should take a `Vec<InternedString>` (or even better, a `Vec<Name>`).

jseyfried commented on src/libsyntax_ext/deriving/custom.rs:
```diff
@@ -12,66 +12,76 @@ use std::panic;
+                                &self.attrs),
+            ItemKind::Enum(ref def, ..) =>
+                mark_attrs_used(def.variants.iter().flat_map(|f| f.node.attrs.iter()).cloned(),
+                                &self.attrs),
```
If we use a visitor, these call should be able to be replaced with `MarkUsed(&self.attrs).visit_item(item_ref)`.

jseyfried commented on src/libsyntax_ext/deriving/custom.rs:
```diff
@@ -88,12 +98,9 @@ impl MultiItemModifier for CustomDerive {
-        // macros, everything is just parsed as a string. Reassign all spans to
-        // the input `item` for better errors here.
-        item.into_iter().flat_map(|item| {
-            ChangeSpan { span: input_span }.fold_item(item)
```
We still want to `ChangeSpan` the new impls generated by custom derives.

nrc commented:
>    Currently this uses a separate attribute named #[proc_macro_attributes(..)], is this the best choice?

What is the motivation for that? I'd prefer a single attribute as proposed by @eddyb.

>    In order to make this work, the check_attribute function had to be modified to not error on attributes marked as used. This is a pretty large change in semantics, is there a better way to do this?

I would prefer not to do this. I think we should keep separate the concepts of used and allowed attributes. White-listing an attribute should make it allowed (scoped to the derive uses, so I guess the white-list has to be per-derive) and mark any uses as used. Other attributes could be marked used by a proc macro, but should still trigger the custom attribute lint.

>    Is switching to MultiItemDecorator the right thing here?

Yes.

keeperofdakeys commented:
So a few things I've done:

- I've renamed KNOWN_ATTRIBUTES to BUILTIN_ATTRIBUTES, and made KNOWN_ATTRIBUTES similar in scope to USED_ATTRIBUTES. CustomDerive then uses attr::mark_known so that the attribute won't trigger a custom attributes error.
- After discussing this on IRC, I've changed back to MultiItemModifier. It has a better call-signature, and CustomDerive may end up having its own copy as a separate trait.
- I've updated the tests a bit, but I still need to write some failing tests for custom attributes errors.

nrc commented:
r? @jseyfried

keeperofdakeys commented:
I've added some basic tests for the proc_macro_derive attribute form with an attributes list, and for  custom attribute errors in deriving items. So unless there are more changes that you'd like me to make, this should be ready to merge.

jseyfried commented on src/libsyntax_ext/proc_macro_registrar.rs:
```diff
@@ -143,38 +145,70 @@ impl<'a> Visitor for CollectCustomDerives<'a> {
+                                       `attributes(foo, bar)`");
+                &[]
+            }).into_iter()
+            .filter_map(|attr| {
```
optional formatting nit:
It would be more idiomatic to either give `})` its own line or have a single line `}).into_iter().filter_map(|attr| {`.

keeperofdakeys commented:
What changes would need to be made to fix https://github.com/rust-lang/rust/pull/37614? Is it as simple as reverting the changes made in https://github.com/rust-lang/rust/pull/36308/commits/3784067edcbcd0614f6c4c88f6445ca17ae27ff6? I've checked the example item errors, and they are successfully reported on the correct item span thanks to this PR.

jseyfried commented:
> What changes would need to be made to fix #37614?

Reverting 3784067, adding a regression test, and editing this PR's initial comment to include "fixes #37614".

keeperofdakeys commented:
I've submitted the required changes to fix https://github.com/rust-lang/rust/issues/37563. This should be ready to merge now.

jseyfried commented:
Excellent, thanks @keeperofdakeys 
@bors r+

bors commented:
:pushpin: Commit 134ef4f has been approved by `jseyfried`

<!-- @bors r=jseyfried 134ef4f7933b87efdc04eac3e8d9a530d56d9cfe -->

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #37192

bors commented:
:pushpin: Commit 134ef4f has been approved by `jseyfried`

<!-- @bors r=jseyfried 134ef4f7933b87efdc04eac3e8d9a530d56d9cfe -->

nikomatsakis commented:
Just saw this; thanks for jumping on this!
